File: src/main/scala/org/xbery/artbeams/users/service/CmsAuthenticationProvider.scala
package org.xbery.artbeams.users.service

import javax.inject.Inject
import org.springframework.security.authentication.{AuthenticationProvider, UsernamePasswordAuthenticationToken}
import org.springframework.security.core.Authentication
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.stereotype.Service

import scala.jdk.CollectionConverters._

/**
  * Spring Security {@link AuthenticationProvider} for logging users into CMS.
  *
  * @author Radek Beran
  */
@Service
class CmsAuthenticationProvider @Inject()(loginService: LoginService) extends AuthenticationProvider {
  import CmsAuthenticationProvider._

  override def authenticate(authentication: Authentication): Authentication = {
    val username = authentication.getName()
    val password = authentication.getCredentials().toString()
    loginService.login(username, password) match {
      case Some(user) =>
        new UsernamePasswordAuthenticationToken(user.id + PrincipalSeparator + user.login, user.password, user.roles.map(role => new SimpleGrantedAuthority(role.name)).asJava)
      case _ =>
        null
    }
  }

  override def supports(authentication: Class[_]): Boolean = {
    authentication.equals(classOf[UsernamePasswordAuthenticationToken])
  }
}

object CmsAuthenticationProvider {
  val PrincipalSeparator = ":"
}

File: src/main/kotlin/org/xbery/artbeams/users/service/CmsAuthenticationProvider.kt
package org.xbery.artbeams.users.service

import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.stereotype.Service

/**
 * Spring Security {@link AuthenticationProvider} for logging users into CMS.
 *
 * @author Radek Beran
 */
@Service
open class CmsAuthenticationProvider(private val loginVerificationService: LoginVerificationService) : AuthenticationProvider {
    override fun authenticate(authentication: Authentication): Authentication? {
        val username = authentication.name
        val password = authentication.credentials.toString()
        val user = loginVerificationService.verifyUser(username, password)
        return if (user != null) {
            loginVerificationService.createAuthenticationToken(user)
        } else {
            null
        }
    }

    override fun supports(authentication: Class<*>): Boolean {
        return authentication == UsernamePasswordAuthenticationToken::class.java
    }
}
